### PR TITLE
cpu/stm32f4: optimized UART driver

### PR DESCRIPTION
**Pull request by haukepetersen**
based on #4114

This driver is now utilizing DMA for transmitting data. This is possible through the changes in the UART driver interface and should save quite some energy as the CPU will now be able to sleep while sending data...

### CODE REVIEW

**Comment by OlegHahm** [comment]
Either the commit set or the title of this PR seems to be off.

**Comment by haukepetersen** [comment]
nope, all fine 

> based on #4114

**Comment by gebart** [comment]
why the change to cpu/stm32f4/include/stm32f407xx.h ? Is that not the vendor-supplied hardware header?

**Comment by haukepetersen** [comment]
yes, it is - but it's shitty... I got used to actually editing those headers instead of introducing x lines of extra code in the drivers...

**Comment by gebart** [comment]
Could you split the stm32f407xx.h change to a separate commit? Makes it easier to get an overview in the future if we want to see what changes we have made to vendor headers.

**Comment by gebart** [comment]
btw, there's a new version of the header available with the Keil MDK (tagged `@version V2.4.0, @date    14-August-2015`), it has some various changes to the registers, you might want to update while you are messing around in this file.

**Comment by haukepetersen** [comment]
rebased and separated the vendor header changes in its own commit

**Comment by haukepetersen** [comment]
also adapted remaining stm32f4 based boards to the changes

**Comment by haukepetersen** [comment]
fixed typo in RCC define for the `msbiot` and squashed it

**Comment by haukepetersen** [comment]
@gebart: are you good with this PR now?

**Comment by gebart** [comment]
I'm fine, did not test it though.

**Comment by haukepetersen** [comment]
fixed a small typing issue with cpp, Travis should be fine now.

File: cpu/stm32f4/periph/uart.c [hunk 1]
```diff
@@ -22,250 +22,212 @@
...
     dev->CR1 = USART_CR1_UE | USART_CR1_TE | USART_CR1_RE;
-
+    /* configure the DMA stream for transmission */
+    dma_poweron(uart_config[uart].dma_stream);
+    stream = dma_stream(uart_config[uart].dma_stream);
+    stream->CR = ((uart_config[uart].dma_chan << 25) |
+                  DMA_SxCR_PL_0 |
+                  DMA_SxCR_MINC |
+                  DMA_SxCR_DIR_0 |
+                  // DMA_SxCR_PFCTRL |
```
Comment (thomaseichinger) [inline cpu/stm32f4/periph/uart.c]: Please remove if not needed.

**Comment by thomaseichinger** [comment]
Could you please also adopt the `nucleo-f401` board?

File: cpu/stm32f4/periph/uart.c (hunk 1 above)
Comment (haukepetersen) [inline cpu/stm32f4/periph/uart.c]: ups :-)

**Comment by haukepetersen** [comment]
done

**Comment by haukepetersen** [comment]
fixed minor comment by @thomaseichinger and squashed. @thomaseichinger: do you ack? There is a ready to use board in our office :-)

**Comment by thomaseichinger** [comment]
ACK & go

**Pull request merged**